rust/rbac-registration/src/cardano/cip509/utils/cip134.rs

```diff
@@ -0,0 +1,111 @@
+//! Utility functions for CIP-0134 address.
+
+use anyhow::{anyhow, Context, Result};
+use pallas::ledger::addresses::Address;
+
+/// Parses CIP-0134 URI and returns an address.
+///
+/// # Errors
+/// - Invalid URI.
+///
+/// # Examples
+///
+/// ```
+/// use pallas::ledger::addresses::{Address, Network};
+/// use rbac_registration::cardano::cip509::utils::parse_cip0134_uri;
+///
+/// let uri = "web+cardano://addr/stake1uyehkck0lajq8gr28t9uxnuvgcqrc6070x3k9r8048z8y5gh6ffgw";
+/// let Address::Stake(address) = parse_cip0134_uri(uri).unwrap() else {
+///     panic!("Unexpected address type");
+/// };
+/// assert_eq!(address.network(), Network::Mainnet);
+/// ```
+pub fn parse_cip0134_uri(uri: &str) -> Result<Address> {
```

Comment:
I like the seperation here.
What if we extend this a little, make a `Cip0134Uri` URI type.
It can use `try_from` for &str to parse a URI into the struct.
`display` can generate the URI back again.
It can then have a method to return the `Address` contained in the URI type.

This way, we preserve the URI itself, and its still just as usable?

Reply:
I would also like to extend this base type with a collection type as well.

`Cip0134UriList` which is just a list of `Cip0134Uri` types, but has a `try_from` method that can create it from a `der_sequence` like the code below does, but it creates a list of all addresses, we would then have methods on the type to return the sequence filtered by `stake` or `addr` or `all`.  That could be three functions, or one function with options, depends on what you think is more ergonomic.

But its able to be iterated on this filtered set of data, because thats how we use it.
I would like the iterator to return the `Cip0134Uri` types, so the logic using it can decide what it wants to do with the address, use it in its text form, or pull the parsed address, etc.

This should be a second ticket on this PR.

The reason for this is we have the `parse_der_sequence` and following logic replicated in multiple places, and its not reusable as is.  This would allow us to have a single set of code which knows how to extract these addresses from a der_sequence we find inside an x509 cert, and not have to re-implement it all the time.

Reply:
I will address the second comment (https://github.com/input-output-hk/catalyst-libs/issues/104 issue) in a separate pull request. The first one should be addressed now.